### ISSUE TITLE
Huge memory consumption with linearSVM associated with large number of classes

### ISSUE THREAD

**Issue by osm3** [topepo/caret#440]
Hi!
Why this code with matrix 1000x4000 consumes 4Gb RAM, and stoppes with out-of memory error after 5 minutes? After stop memory is not released.
library(caret)
MaxLines <- 1000
MaxDims <- 4000
MaxClasses <-  200
set.seed(100)
x <- matrix(,nrow=MaxLines,ncol=MaxDims)
x[,1:MaxDims] <- round(rnorm(MaxDims))
y <- as.factor(round(runif(MaxLines)*(MaxClasses-1)))
control <- trainControl(method="none")
modelSvm <- train(x=x, y=y, method="svmLinear", metric="Accuracy", trControl=control)
i need to use matrix with 150000 X 4000 and number of classes is 500. How can i do it?

**Comment by zachmayer** [comment]
Svms scale quadraticly in terms of ram usage with the dataset size.

150k rows is pretty big for an svm.  You could try an approximate svm.

Not sure why it's using a lot of ram on 1000 rows though.  4K rows is pretty wide-- maybe that's the issue?  Try running a kernlab model outside Caret and see how much ram it uses.

**Comment by zachmayer** [comment]
Woaaaah!  Just saw 500 classes.  That's your problem right there.

**Comment by osm3** [comment]
yes, number of classes is very significant for memory, but it is real and not issue. In my example - 200.

**Comment by topepo** [comment]
Do you have the same problem when you run the code directly from the package? 

Also, try `method="svmLinear2"` which uses the `e1071` package instead of `kernlab`

**Comment by osm3** [comment]
what mean "directly from the package"?
modelSvm <- train(x=x, y=y, method="svmLinear2", metric="Accuracy", trControl=control, tuneGrid = data.frame("cost" = 1))
executed successfully on 500 classes, but memory usage - 1.5GB, and memory did not released after execution.
how can i use SGD for training SVM?

**Comment by zachmayer** [comment]
caret doesn't support SGD for SVM.  I suggest trying [shogun](http://www.shogun-toolbox.org/static/notebook/current/multiclass_reduction.html) or [liblinear](http://scikit-learn.org/stable/modules/generated/sklearn.svm.LinearSVC.html) for that.

The `svmLinear` method comes from the [kernlab](https://cran.r-project.org/web/packages/kernlab/index.html) package.  You can read the docs there to find out how to run a single, linear svm (I don't know how to).

**Comment by zachmayer** [comment]
@topepo It actually might be a good idea to add a `svmLinear3` using liblinear: https://cran.r-project.org/web/packages/LiblineaR/LiblineaR.pdf

That might scale better than kernlab and e1071.

**Comment by topepo** [comment]
@zachmayer I'll add it to the list; they will probably do better in this instance. I think that `kernlab` does particularly poorly. The data that I'll use for useR is fairly big and I let the `e1071` code run for about a week before I killed the process. 

@osm3 `train` is just a wrapper around other packages. To see if it is a `caret` issue or a `kernlab`/`e1071` issue,you should run it using the source package. Taking a look at `getModelInfo("svmLinear", FALSE)[[1]]$fit`, you would fit:

```r
library(kernlab)
out <- ksvm(x = as.matrix(x), y = y,
            kernel = vanilladot(),
            C = param$C)
```

for some value of `C`. You can do the same with `getModelInfo("svmLinear2", FALSE)[[1]]$fit` for `e1071`
